manticore/ethereum/__init__.py

```diff
@@ -931,6 +963,9 @@ def _transaction(self, sort, caller, value=0, address=None, data=None, gaslimit=
             if '_pending_transaction' in state.context:
                 raise EthereumError("This is bad. It should not be a pending transaction")
 
+            if preconstraint is not None:
```

Comment:
this could run into issues if the first four bytes are already constrained to 1 solution, or are effectively concrete, for example if the API is used to execute a function with symbolic arguments

```
contract.func(m.make_symbolic_value())
```

will result in a tx data that is an ArrayProxy (and thus `issymbolic()`), even thought the first four bytes are really concrete. if we add contradicting constraints on the first four bytes (such as a function preconstrain), that will result in an infeasible state, and solver errors raised later.

we can semi-reliably detect if the func has is already constrained by checking if there can be more than 1 solution. something like

```
# ask for 2 solutions. if we only get one back, there can only be 1.
hash_not_constrained = len(state.solve_n(data[:4], 2)) > 1
if hash_not_constrained:
   # hash not constrained, it's safe to do the preconstrain stuff
```

Reply:
The constructed `preconstraint` is a logical disjunction over all possible function selectors (see the `preconstraint = Operators.OR(preconstraint, c)` line),  so I don't think that the implementation could add a contradicting constraint here, unless the concrete function has no entry in the `SolidityMetadata`, which would probably be a programming error.

Invoking the solver here might be bad for performance.

Reply:
I could try to detect simple cases where the selector is concrete without invoking the selector. 

An alternative might be to special-case the `transaction` call here https://github.com/trailofbits/manticore/blob/2a634b707c664f8d0150afa100e280fc55b5408d/manticore/ethereum/account.py#L133-L136

Reply:
> The constructed `preconstraint` is a logical disjunction over all possible function selectors (see the `preconstraint = Operators.OR(preconstraint, c)` line), so I don't think that the implementation could add a contradicting constraint here, unless the concrete function has no entry in the `SolidityMetadata`, which would probably be a programming error

Ah, you're totally right, i retract my comment. when i wrote that i was incorrectly thinking that this preconstraint was constraining it to a specific function. so i guess even if the tx hash bytes were already constrained/concrete, adding this disjunction constraint would technically be redundant, but not do any harm.